{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a87c5910-43b3-4c1e-8227-8630a603a1c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absp /home/gridsan/djuna/.conda/envs/wgs_env/lib/python3.10/site-packages/tabixpy\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import uuid\n",
    "import allel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1cc12bd0-eaa9-4de7-be1a-a3845c8ff072",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = './test_output/DEJ_11898_B01_GRM_WGS_2017-05-15_19.recalibrated_variants.annotated.vcf.gz'\n",
    "path2 = './test_output/DEJ_11898_B01_GRM_WGS_2017-05-15_19.recalibrated_variants.annotated.vcf.gz.tbi'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ee68d3f7-3ce8-4e23-88d6-7336ee785fef",
   "metadata": {},
   "outputs": [],
   "source": [
    "dictionary = np.load('./human_Release_19_GRCh37p13/pos_dictionary.npy', allow_pickle=True).item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "73c48561-a089-4bd6-849a-9eb8136fc2bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/state/partition1/slurm_tmp/21190112.0.0/ipykernel_84177/1477470787.py:1: DtypeWarning: Columns (34,44,46,78,81) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  d = pd.DataFrame(pd.read_table('./test_output/DEJ_11898_B01_GRM_WGS_2017-05-15_19.recalibrated_variants.annotated.coding.txt'))\n"
     ]
    }
   ],
   "source": [
    "d = pd.DataFrame(pd.read_table('./test_output/DEJ_11898_B01_GRM_WGS_2017-05-15_19.recalibrated_variants.annotated.coding.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "8e90b2e9-c44c-486c-b7c5-979075a20d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('log.txt', 'w')\n",
    "\n",
    "path = './test_output/DEJ_11898_B01_GRM_WGS_2017-05-15_19.recalibrated_variants.vcf.gz'\n",
    "path2 = './test_output/DEJ_11898_B01_GRM_WGS_2017-05-15_19.recalibrated_variants.vcf.gz.tbi'\n",
    "callset = allel.read_vcf(path, log = f, fields = '*', region=dictionary['ABCA7'], tabix=path2) ### need to get the sample calls / annotations\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "633d05b4-c2f7-4081-8429-c8897cd4a1ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['samples', 'calldata/AB', 'calldata/AD', 'calldata/DP', 'calldata/GQ', 'calldata/GT', 'calldata/MIN_DP', 'calldata/MQ0', 'calldata/PGT', 'calldata/PID', 'calldata/PL', 'calldata/RGQ', 'calldata/SB', 'variants/AC', 'variants/AF', 'variants/ALT', 'variants/AN', 'variants/BaseQRankSum', 'variants/CHROM', 'variants/DP', 'variants/DS', 'variants/END', 'variants/ExcessHet', 'variants/FILTER_LowQual', 'variants/FILTER_PASS', 'variants/FILTER_VQSRTrancheINDEL99.00to100.00', 'variants/FILTER_VQSRTrancheINDEL99.00to100.00+', 'variants/FILTER_VQSRTrancheSNP99.80to100.00', 'variants/FILTER_VQSRTrancheSNP99.80to100.00+', 'variants/FS', 'variants/HaplotypeScore', 'variants/ID', 'variants/InbreedingCoeff', 'variants/MLEAC', 'variants/MLEAF', 'variants/MQ', 'variants/MQ0', 'variants/MQRankSum', 'variants/NEGATIVE_TRAIN_SITE', 'variants/POS', 'variants/POSITIVE_TRAIN_SITE', 'variants/QD', 'variants/QUAL', 'variants/REF', 'variants/ReadPosRankSum', 'variants/SOR', 'variants/VQSLOD', 'variants/VariantType', 'variants/altlen', 'variants/culprit', 'variants/is_snp', 'variants/numalt'])"
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "e25f5c3e-57b5-4a60-8250-327570c6ae4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[4.181e-04,       nan,       nan],\n",
       "       [4.181e-04,       nan,       nan],\n",
       "       [5.017e-03,       nan,       nan],\n",
       "       ...,\n",
       "       [8.361e-04,       nan,       nan],\n",
       "       [4.181e-04,       nan,       nan],\n",
       "       [7.170e-01,       nan,       nan]], dtype=float32)"
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "e112ad18-30a3-40e0-b837-9f1db0be31c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame(np.concatenate((callset['variants/CHROM'].reshape(-1,1), callset['variants/POS'].reshape(-1,1), \n",
    "               callset['variants/REF'].reshape(-1,1), callset['variants/ALT'],\n",
    "               callset['variants/FILTER_PASS'].reshape(-1,1), callset['variants/numalt'].reshape(-1,1), callset['variants/AF']), axis = 1))\n",
    "df1.columns = ['CHROM', 'POS', 'REF_0', 'ALT_1', 'ALT_2', 'ALT_3', 'FILTER_PASS', 'NUMALT', 'AF_1', 'AF_2', 'AF_3']\n",
    "df2 = (allel.GenotypeArray(callset['calldata/GT']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "348e7f08-80e0-4319-9ce9-b74d137bff45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF_0</th>\n",
       "      <th>ALT_1</th>\n",
       "      <th>ALT_2</th>\n",
       "      <th>ALT_3</th>\n",
       "      <th>FILTER_PASS</th>\n",
       "      <th>NUMALT</th>\n",
       "      <th>AF_1</th>\n",
       "      <th>AF_2</th>\n",
       "      <th>AF_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>1040117</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>1040272</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>1040308</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>1040382</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1040390</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>19</td>\n",
       "      <td>1065197</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>19</td>\n",
       "      <td>1065304</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>19</td>\n",
       "      <td>1065305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>19</td>\n",
       "      <td>1065410</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>19</td>\n",
       "      <td>1065563</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>628 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM      POS REF_0 ALT_1 ALT_2 ALT_3 FILTER_PASS NUMALT      AF_1 AF_2  \\\n",
       "0      19  1040117     A     G                    True      1  0.000418  NaN   \n",
       "1      19  1040272     C     T                    True      1  0.000418  NaN   \n",
       "2      19  1040308     G     A                    True      1  0.005017  NaN   \n",
       "3      19  1040382     G     A                    True      1     0.046  NaN   \n",
       "4      19  1040390     C     T                    True      1  0.000418  NaN   \n",
       "..    ...      ...   ...   ...   ...   ...         ...    ...       ...  ...   \n",
       "623    19  1065197     G     C                    True      1     0.717  NaN   \n",
       "624    19  1065304     C     A                    True      1  0.000418  NaN   \n",
       "625    19  1065305     G     A                    True      1  0.000836  NaN   \n",
       "626    19  1065410     G     C                    True      1  0.000418  NaN   \n",
       "627    19  1065563     G     C                    True      1     0.717  NaN   \n",
       "\n",
       "    AF_3  \n",
       "0    NaN  \n",
       "1    NaN  \n",
       "2    NaN  \n",
       "3    NaN  \n",
       "4    NaN  \n",
       "..   ...  \n",
       "623  NaN  \n",
       "624  NaN  \n",
       "625  NaN  \n",
       "626  NaN  \n",
       "627  NaN  \n",
       "\n",
       "[628 rows x 11 columns]"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ffff6c-b70d-4354-a1e4-a0548f28e778",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging on position: postion to index dictionary "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "244926fa-35c1-4889-b655-1173de241590",
   "metadata": {},
   "outputs": [],
   "source": [
    "dd = pd.merge(df1, d[['POS', 'ID', 'REF', 'ALT', 'EFFECT', 'IMPACT', 'GENE', 'GENEID', 'HGVS_C', 'HGVC_P', 'LOF', 'NMD']], on = 'POS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "9843ca24-660e-46a5-9de1-6d286cc58f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "dd = dd.loc[dd['FILTER_PASS']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "id": "6c5bd83f-532f-40c9-8061-62735dabd40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos = callset['variants/POS']\n",
    "indices = dict(zip(pos, range(len(pos))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "32b7813c-47e9-4d6e-b4e6-8df8e03a8b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "select = [indices[i] for i in dd['POS']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "aa451560-5e59-4712-9314-c08fa157b691",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype = callset['calldata/GT'][select]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "f1de8af1-5df8-407e-bc36-f774c8f8fd91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_genotypes_to_str(genotype):\n",
    "    converted = [str(i[0]) + '/' + str(i[1]) for i in genotype]\n",
    "    return converted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "aa6f1d88-30b0-41e7-8377-687774191a24",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_df = pd.DataFrame([convert_genotypes_to_str(x) for x in genotype])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "71bf4e65-88e6-48ad-8a83-03fec0b3fd20",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_df.columns = callset['samples']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "6f90e2e2-12cf-4386-9fe4-7fd8282d02ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = np.unique(genotypes_df.iloc[0], return_counts = True)\n",
    "temp2 = np.unique(genotypes_df.iloc[1], return_counts = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "fca34206-dd43-4a2c-a852-ca6d4d3bcc5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = np.unique(genotypes_df)\n",
    "\n",
    "frequencies = pd.DataFrame([[np.sum(genotypes_df.iloc[i]==x) for x in g] for i in range(genotypes_df.shape[0])])\n",
    "frequencies.columns = 'N ' + g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "8f5c5f41-72f2-45bd-b9e4-412bd1b8469a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF_0</th>\n",
       "      <th>ALT_1</th>\n",
       "      <th>ALT_2</th>\n",
       "      <th>ALT_3</th>\n",
       "      <th>FILTER_PASS</th>\n",
       "      <th>NUMALT</th>\n",
       "      <th>AF_1</th>\n",
       "      <th>AF_2</th>\n",
       "      <th>...</th>\n",
       "      <th>SM-CTEN7</th>\n",
       "      <th>SM-CTEN8</th>\n",
       "      <th>SM-CTEN9</th>\n",
       "      <th>SM-CTENA</th>\n",
       "      <th>SM-CTENC</th>\n",
       "      <th>SM-CTEND</th>\n",
       "      <th>SM-CTENE</th>\n",
       "      <th>SM-CTENF</th>\n",
       "      <th>SM-CTENG</th>\n",
       "      <th>SM-CTENH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>1041971</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.001672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>1042158</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>1042791</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>1042809</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1043103</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>19</td>\n",
       "      <td>1065018</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.716</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>19</td>\n",
       "      <td>1065145</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>19</td>\n",
       "      <td>1065304</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>19</td>\n",
       "      <td>1065305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>19</td>\n",
       "      <td>1065410</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 1222 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM      POS REF_0 ALT_1 ALT_2 ALT_3 FILTER_PASS NUMALT      AF_1 AF_2  \\\n",
       "0     19  1041971     T     G                    True      1  0.001672  NaN   \n",
       "1     19  1042158     G     C                    True      1  0.000836  NaN   \n",
       "2     19  1042791     G     A                    True      1  0.000418  NaN   \n",
       "3     19  1042809     A     G                    True      1     0.457  NaN   \n",
       "4     19  1043103     G     A                    True      1     0.055  NaN   \n",
       "..   ...      ...   ...   ...   ...   ...         ...    ...       ...  ...   \n",
       "62    19  1065018     G     T                    True      1     0.716  NaN   \n",
       "63    19  1065145     C     A                    True      1  0.000418  NaN   \n",
       "64    19  1065304     C     A                    True      1  0.000418  NaN   \n",
       "65    19  1065305     G     A                    True      1  0.000836  NaN   \n",
       "66    19  1065410     G     C                    True      1  0.000418  NaN   \n",
       "\n",
       "    ... SM-CTEN7 SM-CTEN8 SM-CTEN9 SM-CTENA SM-CTENC SM-CTEND SM-CTENE  \\\n",
       "0   ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "1   ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "2   ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "3   ...      1/1      0/0      0/1      0/0      1/1      0/1      1/1   \n",
       "4   ...      0/0      0/0      0/0      0/0      0/1      0/0      0/0   \n",
       "..  ...      ...      ...      ...      ...      ...      ...      ...   \n",
       "62  ...      1/1      1/1      0/1      0/1      1/1      0/1      1/1   \n",
       "63  ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "64  ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "65  ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "66  ...      0/0      0/0      0/0      0/0      0/0      0/0      0/0   \n",
       "\n",
       "   SM-CTENF SM-CTENG SM-CTENH  \n",
       "0       0/0      0/0      0/0  \n",
       "1       0/0      0/0      0/0  \n",
       "2       0/0      0/0      0/0  \n",
       "3       0/0      0/1      0/1  \n",
       "4       0/0      0/0      0/0  \n",
       "..      ...      ...      ...  \n",
       "62      1/1      0/1      0/1  \n",
       "63      0/0      0/0      0/0  \n",
       "64      0/0      0/0      0/0  \n",
       "65      0/0      0/0      0/0  \n",
       "66      0/0      0/0      0/0  \n",
       "\n",
       "[67 rows x 1222 columns]"
      ]
     },
     "execution_count": 314,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_variants = pd.concat((dd, frequencies, genotypes_df), axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ef829f7-5099-42b3-8dc9-2278d881b4dd",
   "metadata": {},
   "source": [
    "## clean up up until here and write tests\n",
    "## research the QC metrics -> what vars are considered good quality?\n",
    "## read through the pipeline that ROSMAP used\n",
    "## add to main.py --> want the all_variants file above concatenated per gene of interest\n",
    "### prompt: return all the moderate to high coding variants in this set of genes\n",
    "##### then I can use this package to get the tables for my genes of interest; subset by variants of interest, and add columns that indicate how many of the\n",
    "### individuals have been sequenced and their APOE4 status... (think about how best to organize that table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "6cf051e5-5fb7-4cf8-a865-4700598f271c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>1040117</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>1040272</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>1040308</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>1040382</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1040390</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>19</td>\n",
       "      <td>1065197</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>19</td>\n",
       "      <td>1065304</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>19</td>\n",
       "      <td>1065305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>19</td>\n",
       "      <td>1065410</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>19</td>\n",
       "      <td>1065563</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>628 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1  2  3 4  5      6\n",
       "0    19  1040117  A  G        True\n",
       "1    19  1040272  C  T        True\n",
       "2    19  1040308  G  A        True\n",
       "3    19  1040382  G  A        True\n",
       "4    19  1040390  C  T        True\n",
       "..   ..      ... .. .. .. ..   ...\n",
       "623  19  1065197  G  C        True\n",
       "624  19  1065304  C  A        True\n",
       "625  19  1065305  G  A        True\n",
       "626  19  1065410  G  C        True\n",
       "627  19  1065563  G  C        True\n",
       "\n",
       "[628 rows x 7 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 ### add the annotation data to this array \n",
    "### save this array and the df2 as a np array\n",
    "### then use both arrays to subset df2 to get specific variants\n",
    "### summarize how many of these individuals we have and how many are sequenced alreaady\n",
    "### save also as csv to make per individual plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "87b3ea39-ecdb-4406-9f97-17eb7cbc0f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "dd = pd.merge(df1, d[['CHROM', 'POS', 'ID', 'REF', 'ALT', 'EFFECT', 'IMPACT', 'GENE', 'GENEID', 'HGVS_C', 'HGVC_P', 'LOF', 'NMD']], left_on = 1, right_on = 'POS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "ac6ef821-a7d9-4661-bf7b-ca2025aa463f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>EFFECT</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>GENEID</th>\n",
       "      <th>HGVS_C</th>\n",
       "      <th>HGVC_P</th>\n",
       "      <th>LOF</th>\n",
       "      <th>NMD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>19</td>\n",
       "      <td>1045035</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>1045035</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ENSG00000064687</td>\n",
       "      <td>c.1250C&gt;A</td>\n",
       "      <td>p.Ser417*</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>19</td>\n",
       "      <td>1053362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>1053362</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ENSG00000064687</td>\n",
       "      <td>c.3255G&gt;A</td>\n",
       "      <td>p.Trp1085*</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>19</td>\n",
       "      <td>1054255</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>1054255</td>\n",
       "      <td>rs201060968</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ENSG00000064687</td>\n",
       "      <td>c.3641G&gt;A</td>\n",
       "      <td>p.Trp1214*</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>19</td>\n",
       "      <td>1058685</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>1058685</td>\n",
       "      <td>rs146448899</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>ENSG00000064687</td>\n",
       "      <td>c.5218C&gt;T</td>\n",
       "      <td>p.Gln1740*</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "      <td>(ABCA7|ENSG00000064687|1|1.00)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0        1  2  3 4 5     6  CHROM      POS           ID REF ALT  \\\n",
       "11  19  1045035  C  A      True     19  1045035            .   C   A   \n",
       "28  19  1053362  G  A      True     19  1053362            .   G   A   \n",
       "33  19  1054255  G  A      True     19  1054255  rs201060968   G   A   \n",
       "55  19  1058685  C  T      True     19  1058685  rs146448899   C   T   \n",
       "\n",
       "         EFFECT IMPACT   GENE           GENEID     HGVS_C      HGVC_P  \\\n",
       "11  stop_gained   HIGH  ABCA7  ENSG00000064687  c.1250C>A   p.Ser417*   \n",
       "28  stop_gained   HIGH  ABCA7  ENSG00000064687  c.3255G>A  p.Trp1085*   \n",
       "33  stop_gained   HIGH  ABCA7  ENSG00000064687  c.3641G>A  p.Trp1214*   \n",
       "55  stop_gained   HIGH  ABCA7  ENSG00000064687  c.5218C>T  p.Gln1740*   \n",
       "\n",
       "                               LOF                             NMD  \n",
       "11  (ABCA7|ENSG00000064687|1|1.00)  (ABCA7|ENSG00000064687|1|1.00)  \n",
       "28  (ABCA7|ENSG00000064687|1|1.00)  (ABCA7|ENSG00000064687|1|1.00)  \n",
       "33  (ABCA7|ENSG00000064687|1|1.00)  (ABCA7|ENSG00000064687|1|1.00)  \n",
       "55  (ABCA7|ENSG00000064687|1|1.00)  (ABCA7|ENSG00000064687|1|1.00)  "
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd.loc[np.array(dd['NMD'])== '(ABCA7|ENSG00000064687|1|1.00)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "c7c6f809-6af5-4b87-b0d1-a30dc1c0000f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['frameshift_variant', 'frameshift_variant&splice_region_variant',\n",
       "       'missense_variant', 'missense_variant&splice_region_variant',\n",
       "       'splice_acceptor_variant&intron_variant',\n",
       "       'splice_donor_variant&intron_variant', 'stop_gained'], dtype=object)"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dd['EFFECT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "fd53cebf-b811-4443-b235-9ec191b79c15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['', '*', 'C'], dtype=object)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dd.loc[:,4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "387655e8-d2b4-4799-a529-41477ab244e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeArray shape=(628, 1196, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">1191</th><th style=\"text-align: center\">1192</th><th style=\"text-align: center\">1193</th><th style=\"text-align: center\">1194</th><th style=\"text-align: center\">1195</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"12\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">625</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">626</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">627</th><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">0/1</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeArray shape=(628, 1196, 2) dtype=int8>\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "...\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "1/1 0/1 1/1 0/1 0/1 ... 0/1 1/1 1/1 0/1 0/1"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat((df1, df2), axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "fdf48382-2fc4-428e-a194-f6f69fcca414",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True, False,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True, False,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "       False,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True, False,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True, False,  True,  True,  True,  True,  True,\n",
       "        True,  True, False,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True, False,\n",
       "       False, False,  True, False,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True, False, False, False, False,\n",
       "        True,  True, False, False,  True,  True, False,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True, False,  True,  True, False,  True,\n",
       "        True,  True,  True,  True,  True, False,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True, False,  True,  True, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False,  True, False,  True, False,\n",
       "       False, False, False,  True,  True, False, False,  True,  True,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True, False,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True, False,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "       False, False,  True,  True,  True,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False,  True,  True,  True,\n",
       "        True, False,  True,  True,  True,  True,  True, False,  True,\n",
       "       False, False, False, False, False,  True, False,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True])"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abbce4e2-e1f5-4fad-a6c3-3288ebc1027f",
   "metadata": {},
   "outputs": [],
   "source": [
    "“CHROM”, “POS”, “ID”, “REF”, “ALT”, “QUAL”, “FILTER”, “INFO” and “FORMAT”"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "22a38190-50a0-4580-98cc-2e6cecb872aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>1040117</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>357.480011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>1040272</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>391.480011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>1040308</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5317.850098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>1040382</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>48615.78125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1040390</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>271.480011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>19</td>\n",
       "      <td>1065197</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>921820.125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>19</td>\n",
       "      <td>1065304</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>162.479996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>19</td>\n",
       "      <td>1065305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1239.709961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>19</td>\n",
       "      <td>1065410</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>382.480011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>19</td>\n",
       "      <td>1065563</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1018808.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>628 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1  2  3 4  5             6\n",
       "0    19  1040117  A  G         357.480011\n",
       "1    19  1040272  C  T         391.480011\n",
       "2    19  1040308  G  A        5317.850098\n",
       "3    19  1040382  G  A        48615.78125\n",
       "4    19  1040390  C  T         271.480011\n",
       "..   ..      ... .. .. .. ..          ...\n",
       "623  19  1065197  G  C         921820.125\n",
       "624  19  1065304  C  A         162.479996\n",
       "625  19  1065305  G  A        1239.709961\n",
       "626  19  1065410  G  C         382.480011\n",
       "627  19  1065563  G  C          1018808.5\n",
       "\n",
       "[628 rows x 7 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "06a8953e-8709-4015-8c63-ce03979bf249",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'variants/INFO'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [51]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m np\u001b[38;5;241m.\u001b[39munique(\u001b[43mcallset\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mvariants/INFO\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'variants/INFO'"
     ]
    }
   ],
   "source": [
    "np.unique(callset['variants/INFO'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2ff687b3-5dd5-4f69-906a-764965fab412",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', 'CTTTTTTT', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', 'TAAA', '', '', '', '', '', '', '', '', '', '',\n",
       "       'CAAA', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', 'C', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '*', '', '*', '', '*',\n",
       "       '', '*', '*', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', 'A', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', 'A', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', 'A', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '',\n",
       "       'GAGGGAGGGAAAGGGGGAGGAGAAGGGGGAGAAGTAGGAGGAAGAGGAGGGGGAGGAGAAGGGGGAGGAGTAGGAGGGAGAGGAGGAGGAGGGA',\n",
       "       '', '', '*', '', '*', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', 'C', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', 'ATTTAT', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', 'AT', 'TATATA', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', 'C', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '*', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', '', '', '', '*', '',\n",
       "       '*', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '',\n",
       "       '', '', '', '', '', '', '', '', '', '', '', ''], dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset['variants/ALT'][:,2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3633a7ad-82fc-4891-8fbe-2a526d7ed696",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A', 'C', 'G', 'G', 'C', 'C', 'A', 'A', 'A', 'C', 'A', 'A', 'G',\n",
       "       'G', 'G', 'G', 'A', 'A', 'C', 'G', 'G', 'T', 'A', 'A', 'C', 'G',\n",
       "       'C', 'T', 'G', 'CCT', 'C', 'G', 'C', 'T', 'AAT', 'G', 'C', 'C',\n",
       "       'G', 'A', 'C', 'C', 'G', 'G', 'G', 'G', 'T', 'G', 'T', 'C', 'C',\n",
       "       'A', 'A', 'G', 'C', 'C', 'C', 'C', 'A', 'G', 'G', 'G', 'C', 'C',\n",
       "       'CT', 'C', 'G', 'G', 'G', 'C', 'C', 'A', 'T', 'C', 'CTTTTT', 'T',\n",
       "       'G', 'C', 'C', 'C', 'C', 'C', 'A', 'T', 'G', 'C', 'C', 'G', 'C',\n",
       "       'C', 'G', 'C', 'C', 'G', 'G', 'A', 'C', 'C', 'T', 'C', 'G', 'T',\n",
       "       'A', 'A', 'T', 'C', 'C', 'C', 'G', 'T', 'G', 'G', 'T', 'A', 'C',\n",
       "       'G', 'TCCAG', 'C', 'G', 'C', 'C', 'C', 'C', 'T', 'C', 'T', 'TG',\n",
       "       'A', 'G', 'GCTGCGGGACACCATGCGCGCCATGGGGCTCAGCCGCGCGGTGCT', 'C',\n",
       "       'A', 'G', 'A', 'C', 'C', 'A', 'C', 'C', 'G', 'G', 'AGGAGCAG', 'A',\n",
       "       'G', 'C', 'G', 'T', 'G', 'C', 'T', 'C', 'A', 'AGCTGGGCGTGGTGGCAC',\n",
       "       'G', 'G', 'G', 'TAA', 'A', 'G', 'G', 'T', 'G', 'CA', 'T', 'G', 'A',\n",
       "       'A', 'C', 'AAAAAC', 'AAAAC', 'AAAC', 'AAC', 'AC', 'CAA', 'A', 'C',\n",
       "       'G', 'G', 'C', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'G', 'C',\n",
       "       'CA', 'A', 'AC', 'C', 'A', 'C', 'G', 'G', 'C', 'A', 'T', 'C', 'C',\n",
       "       'G', 'C', 'C', 'G', 'A', 'T', 'TCCCCGTGAGCCCCCCCACTCCCA',\n",
       "       'AGCCCCCCC',\n",
       "       'GCCCCCCCACTCCCACCCCGTGAGCCCCCCCACCACTCCCTCCCCGTGAGCCCCCCACCACTCCCTCCCCGTGAACCCCCCACCACTCCCTCCCCGTGAGCCCCCCCACCACTCCCTCCCCGTGAGCCCCCCACCACTCCCTCCCCGTGAACCCCCCACCACTCCCTCCCCGTGAGCCCCCCCACCACTCCCTCCCCGTGAA',\n",
       "       'C', 'CCCACTCCCA', 'CTCCCACCCCGTGAG',\n",
       "       'ACCCCGTGAGCCCCCCCACCACTCCCTCCCCGTGAGCCCCCCACCACTCCCTCCCCGTGAACCCCCCACCACTCCCTCCCCGTGAGCCCCCCCACCACTCCCTCCCCGTGAGCCCCCCACCACTCCCTCCCCGTGAACCCCCCACCACTCCCTCCCCGTGAGCCCCCCCACCACTCCCTCCCCGTGAACCCCCCACCACTCCCT',\n",
       "       'GTGAGCC', 'C', 'C', 'C', 'C', 'A', 'G', 'C', 'G', 'GC', 'C', 'C',\n",
       "       'G', 'C', 'GT', 'T', 'A', 'T', 'T', 'C', 'A', 'G', 'A', 'C', 'C',\n",
       "       'T', 'A', 'G', 'C', 'C', 'C', 'G', 'C', 'G', 'G', 'T', 'C', 'A',\n",
       "       'G', 'T', 'A', 'A', 'A', 'A',\n",
       "       'AACAAATAAATAAAAATAAGGCTATCACAGGGTTTTGGATGTAGCATTTAATAGGTGCTCTGT',\n",
       "       'T', 'G', 'C', 'G', 'A', 'C', 'A', 'C', 'G', 'G', 'C', 'C', 'A',\n",
       "       'C', 'C', 'C', 'AAATAATAATAATAAT', 'A', 'T', 'A', 'T', 'A', 'T',\n",
       "       'CAG', 'A', 'G', 'G', 'G', 'G', 'CTG', 'G', 'A', 'C', 'C',\n",
       "       'TACTGAGGTCCACGTGGGTAGGCAACCTTGCCCAAGGCTGGGTGTGACAC', 'A', 'G',\n",
       "       'G', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'C', 'GGGA', 'T', 'G',\n",
       "       'G', 'G', 'AAGAGGTGGAAGGGGAGGCGAAGGG', 'A', 'A', 'AGGGG', 'AGGGAG',\n",
       "       'G',\n",
       "       'GGGAGGAGAAGGGGGGAGGAGGAGGGGAAGGGAGAGGAGGAGGGGGAGAAGAAGGGGAGGAGTAGGAGGGGAGCAGAAGGGGAGGAATAGGAGGGGGAGGAGGAGGGGGAGGGAAAGGGGGAGGAGAAGGGGGAGAAGTAGGAGGAAGAGGAGGGGGAGGAGAAGGGGGAGGAGTAGGAGGGAGAGGAGGAGGAGGGAGAGGAGGGGGAGGAGAAGGAGAGGAGGAGGA',\n",
       "       'AAGGGGGG', 'AGGGGG', 'G', 'GGA', 'AAG', 'G', 'G', 'AG', 'A',\n",
       "       'AGG', 'G', 'G', 'GTAGGA', 'AGGAGGGGAGCAGAAGGG', 'GGGAGCAGAA', 'G',\n",
       "       'C', 'A', 'A', 'G', 'A', 'T', 'AGG', 'G', 'GA', 'G', 'G', 'G', 'G',\n",
       "       'GAGGAGGAGGGGGAGGGAA', 'G', 'G', 'G',\n",
       "       'AGGGGGAGGGAAAGGGGGAGGAGAAGGGGGAGAAGTAGGAGGAAGAGGAGGGGGAGGAGAAGGGGGAGGAGTAG',\n",
       "       'GGGGAGGGAAAGGGGGAGGAGAAGGGGGAGAAGTAGGAGGAAGAGGAGGGGGAGGAGAAGGGGGAGGAGTAGGAGGGAGAGGAGGAGGAGGGA',\n",
       "       'GAGGGAA', 'AGGGAAAG', 'G', 'AAAGGGG', 'A', 'G', 'AG', 'A', 'A',\n",
       "       'A', 'G', 'TAGGAGGGAGAGGAGG', 'G', 'GGAGAAGGA', 'C', 'G', 'A', 'C',\n",
       "       'C', 'G', 'T', 'C', 'C', 'C', 'C', 'G', 'C', 'C', 'G', 'T', 'A',\n",
       "       'G', 'CT', 'C', 'G', 'C', 'A', 'G', 'A', 'C', 'A', 'G', 'G', 'C',\n",
       "       'T', 'T', 'C', 'G', 'C', 'G', 'C', 'G', 'G', 'G', 'T', 'G', 'CTCG',\n",
       "       'G', 'CTT', 'C', 'C', 'C', 'T', 'C', 'CT', 'C', 'GC', 'C', 'A',\n",
       "       'C', 'G', 'C', 'G', 'C', 'T', 'C', 'G', 'G', 'G', 'T', 'A', 'A',\n",
       "       'A', 'C', 'A', 'G', 'G', 'GTGCGGGGGCTGCT', 'C', 'A', 'G', 'G', 'C',\n",
       "       'C', 'T', 'C', 'G', 'TAGAG', 'A', 'GAGAA', 'GA', 'A', 'A', 'G',\n",
       "       'C', 'A', 'G', 'C', 'C', 'C', 'G', 'G', 'A', 'G', 'G', 'C', 'T',\n",
       "       'T', 'A', 'A', 'C', 'G', 'A', 'G', 'C', 'T', 'C', 'T', 'G', 'A',\n",
       "       'C', 'C', 'A', 'G', 'A', 'G', 'A', 'C', 'G', 'G', 'A', 'G', 'A',\n",
       "       'T', 'T', 'T', 'T', 'TTC', 'TC', 'C', 'C', 'T', 'T', 'CT', 'T',\n",
       "       'G', 'C', 'G', 'G', 'G', 'T', 'C', 'G', 'T', 'G', 'CAT', 'A', 'G',\n",
       "       'C', 'A', 'C', 'A', 'T', 'C', 'C', 'G', 'C', 'G', 'G', 'C', 'G',\n",
       "       'G', 'CA', 'A', 'G', 'C', 'A', 'T', 'G', 'C', 'CAA', 'C', 'T', 'C',\n",
       "       'G', 'C', 'A', 'A', 'C', 'T', 'G', 'T', 'T', 'C', 'T', 'C', 'T',\n",
       "       'C', 'G', 'T', 'G', 'C', 'C', 'A', 'T', 'A', 'C', 'T', 'C', 'G',\n",
       "       'C', 'C', 'GC', 'C', 'TGGCTCCACCCACACCATGGCCCCGCCCCATACTCA', 'G',\n",
       "       'GCCCCATACTCATGCTGGCTCCACCCACACCATGGCCCCA', 'C', 'CTGG', 'G', 'C',\n",
       "       'C', 'A', 'C', 'C', 'C', 'A', 'ACCAGTGTGGCCCCGC',\n",
       "       'TACTCATGCTGTCTCCACCCACACCATGGCCCCGCCCCATACTCATGCTGGCTCCACCCACACTATGGCCCTGCCCCACACCCATCCCAGCTCCACCCACACCATGGCCCCGCCCCATACTCATGCTGGCTCCACCCAC',\n",
       "       'TGCCCCACACCCATCCCAGCTCCACCCACACCATGGCCCC', 'C', 'C', 'C', 'C',\n",
       "       'A', 'C', 'C', 'A', 'G', 'G', 'G', 'C', 'A', 'C', 'G', 'C', 'G',\n",
       "       'G', 'G', 'C', 'C', 'A', 'G', 'G', 'C', 'G', 'G', 'C', 'C', 'G',\n",
       "       'C', 'G', 'G', 'G'], dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset['variants/REF']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "47525acd-39f0-414b-93c5-9d2fbb7ca3bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('log.txt', 'w')\n",
    "callset = allel.read_vcf(path, log = f, fields = '*', region=dictionary['ABCA7'], tabix=path2)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "e180e111-56b9-4658-9e78-349f881002e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['variants/1000Gp3_AA_GF',\n",
       " 'variants/1000Gp3_AF',\n",
       " 'variants/1000Gp3_HomC',\n",
       " 'variants/1000Gp3_RA_GF',\n",
       " 'variants/1000Gp3_RR_GF',\n",
       " 'variants/AC',\n",
       " 'variants/ACMG_GENE',\n",
       " 'variants/ACMG_INHRT',\n",
       " 'variants/ACMG_MIM_GENE',\n",
       " 'variants/ACMG_PATH',\n",
       " 'variants/AF',\n",
       " 'variants/ALT',\n",
       " 'variants/AN',\n",
       " 'variants/ANN',\n",
       " 'variants/AR_GENE',\n",
       " 'variants/BaseQRankSum',\n",
       " 'variants/CADD_phred',\n",
       " 'variants/CHROM',\n",
       " 'variants/CLNDBN',\n",
       " 'variants/CLNDSDB',\n",
       " 'variants/CLNDSDBID',\n",
       " 'variants/CLNSIG',\n",
       " 'variants/COSMIC_CNT',\n",
       " 'variants/DNA',\n",
       " 'variants/DP',\n",
       " 'variants/DS',\n",
       " 'variants/END',\n",
       " 'variants/Entrez_gene_id',\n",
       " 'variants/Essential_gene',\n",
       " 'variants/ExAC_AF',\n",
       " 'variants/ExcessHet',\n",
       " 'variants/FILTER_LowQual',\n",
       " 'variants/FILTER_PASS',\n",
       " 'variants/FILTER_VQSRTrancheINDEL99.00to100.00',\n",
       " 'variants/FILTER_VQSRTrancheINDEL99.00to100.00+',\n",
       " 'variants/FILTER_VQSRTrancheSNP99.80to100.00',\n",
       " 'variants/FILTER_VQSRTrancheSNP99.80to100.00+',\n",
       " 'variants/FS',\n",
       " 'variants/GDI',\n",
       " 'variants/GDI-Phred',\n",
       " 'variants/GERP++_NR',\n",
       " 'variants/GERP++_RS',\n",
       " 'variants/HaplotypeScore',\n",
       " 'variants/ID',\n",
       " 'variants/InbreedingCoeff',\n",
       " 'variants/LCR_hs37d5',\n",
       " 'variants/LINE',\n",
       " 'variants/LOF',\n",
       " 'variants/LTR',\n",
       " 'variants/LoFtool_score',\n",
       " 'variants/Low_complexity',\n",
       " 'variants/MGI_mouse_gene',\n",
       " 'variants/MLEAC',\n",
       " 'variants/MLEAF',\n",
       " 'variants/MQ',\n",
       " 'variants/MQ0',\n",
       " 'variants/MQRankSum',\n",
       " 'variants/MutationAssessor_pred',\n",
       " 'variants/MutationTaster_pred',\n",
       " 'variants/NEGATIVE_TRAIN_SITE',\n",
       " 'variants/NMD',\n",
       " 'variants/OMIM_MENDELIAN_DELDUP_ID',\n",
       " 'variants/OMIM_MENDELIAN_DELDUP_INHRT',\n",
       " 'variants/OMIM_MENDELIAN_DELDUP_PHENO',\n",
       " 'variants/OMIM_MENDELIAN_MUT_ID',\n",
       " 'variants/OMIM_MENDELIAN_MUT_INHRT',\n",
       " 'variants/OMIM_MENDELIAN_MUT_PHENO',\n",
       " 'variants/OMIM_NONDIS_DELDUP_ID',\n",
       " 'variants/OMIM_NONDIS_DELDUP_INHRT',\n",
       " 'variants/OMIM_NONDIS_DELDUP_PHENO',\n",
       " 'variants/OMIM_NONDIS_MUT_ID',\n",
       " 'variants/OMIM_NONDIS_MUT_INHRT',\n",
       " 'variants/OMIM_NONDIS_MUT_PHENO',\n",
       " 'variants/OMIM_SUSCEPT_DELDUP_ID',\n",
       " 'variants/OMIM_SUSCEPT_DELDUP_INHRT',\n",
       " 'variants/OMIM_SUSCEPT_DELDUP_PHENO',\n",
       " 'variants/OMIM_SUSCEPT_MUT_ID',\n",
       " 'variants/OMIM_SUSCEPT_MUT_INHRT',\n",
       " 'variants/OMIM_SUSCEPT_MUT_PHENO',\n",
       " 'variants/POS',\n",
       " 'variants/POSITIVE_TRAIN_SITE',\n",
       " 'variants/Pathway(BioCarta)_short',\n",
       " 'variants/Polyphen2_HDIV_pred',\n",
       " 'variants/Polyphen2_HVAR_pred',\n",
       " 'variants/QD',\n",
       " 'variants/QUAL',\n",
       " 'variants/RC',\n",
       " 'variants/REF',\n",
       " 'variants/RNA',\n",
       " 'variants/RVIS',\n",
       " 'variants/RVIS_percentile',\n",
       " 'variants/ReadPosRankSum',\n",
       " 'variants/Regulome_dbSNP141',\n",
       " 'variants/SIFT_pred',\n",
       " 'variants/SINE',\n",
       " 'variants/SOR',\n",
       " 'variants/Satellite',\n",
       " 'variants/Simple_repeats',\n",
       " 'variants/Uniprot_aapos_Polyphen2',\n",
       " 'variants/Uniprot_id_Polyphen2',\n",
       " 'variants/VQSLOD',\n",
       " 'variants/VariantType',\n",
       " 'variants/ZFIN_zebrafish_gene',\n",
       " 'variants/ZFIN_zebrafish_phenotype_tag',\n",
       " 'variants/altlen',\n",
       " 'variants/culprit',\n",
       " 'variants/dbSNPBuildID',\n",
       " 'variants/is_snp',\n",
       " 'variants/numalt',\n",
       " 'variants/phastCons20way_mammalian',\n",
       " 'variants/phyloP20way_mammalian']"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(callset.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "4e9308f5-9a05-455c-8363-017468bbe39a",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../human_Release_19_GRCh37p13/chr_dictionary.npy'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [187]\u001b[0m, in \u001b[0;36m<cell line: 14>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m gencode_genes[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlocation\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mstr\u001b[39m(x[\u001b[38;5;241m0\u001b[39m]) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(x[\u001b[38;5;241m1\u001b[39m]) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(x[\u001b[38;5;241m2\u001b[39m]) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m sele]\n\u001b[1;32m     12\u001b[0m position_dictionary \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\u001b[38;5;28mzip\u001b[39m(gencode_genes[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m'\u001b[39m], gencode_genes[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlocation\u001b[39m\u001b[38;5;124m'\u001b[39m]))\n\u001b[0;32m---> 14\u001b[0m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m../human_Release_19_GRCh37p13/chr_dictionary.npy\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mchromosome_dictionary\u001b[49m\u001b[43m)\u001b[49m \n\u001b[1;32m     15\u001b[0m np\u001b[38;5;241m.\u001b[39msave(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../human_Release_19_GRCh37p13/pos_dictionary.npy\u001b[39m\u001b[38;5;124m'\u001b[39m, position_dictionary)\n",
      "File \u001b[0;32m<__array_function__ internals>:180\u001b[0m, in \u001b[0;36msave\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/wgs_env/lib/python3.10/site-packages/numpy/lib/npyio.py:498\u001b[0m, in \u001b[0;36msave\u001b[0;34m(file, arr, allow_pickle, fix_imports)\u001b[0m\n\u001b[1;32m    496\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m file\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.npy\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    497\u001b[0m         file \u001b[38;5;241m=\u001b[39m file \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.npy\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m--> 498\u001b[0m     file_ctx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mwb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m file_ctx \u001b[38;5;28;01mas\u001b[39;00m fid:\n\u001b[1;32m    501\u001b[0m     arr \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masanyarray(arr)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../human_Release_19_GRCh37p13/chr_dictionary.npy'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "gencode = pd.read_table('./human_Release_19_GRCh37p13/gencode.v19.annotation.gff3.gz', comment=\"#\", sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start' , 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "gencode_genes = gencode[(gencode.feature == \"gene\")][['seqname', 'start', 'end', 'attribute']].copy().reset_index().drop('index', axis=1)\n",
    "gencode_genes['names'] = gencode_genes['attribute'].str.split(';',expand=True).loc[:,5].str.split('=', expand=True).loc[:,1]\n",
    "gencode_genes['chr'] = [x.replace('chr','') for x in gencode_genes['seqname']]\n",
    "\n",
    "chromosome_dictionary = dict(zip(gencode_genes['names'], gencode_genes['seqname']))\n",
    "\n",
    "sele = np.array(gencode_genes[['chr', 'start','end']])\n",
    "gencode_genes['location'] = [str(x[0]) + ':' + str(x[1]) + '-' + str(x[2]) for x in sele]\n",
    "position_dictionary = dict(zip(gencode_genes['names'], gencode_genes['location']))\n",
    "\n",
    "np.save('../human_Release_19_GRCh37p13/chr_dictionary.npy', chromosome_dictionary) \n",
    "np.save('./human_Release_19_GRCh37p13/pos_dictionary.npy', position_dictionary) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "9fc20364-6ab6-47aa-a622-5dd81ded78cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('./human_Release_19_GRCh37p13/pos_dictionary.npy', position_dictionary) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "ff7c5b0c-a838-404b-b7d5-d041aecb0122",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "e4d32bf3-75b9-4db8-a525-359bfb170774",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'19:1040102-1065571'"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position_dictionary['ABCA7']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wgs_env",
   "language": "python",
   "name": "wgs_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
